### PR TITLE
feat: add time input

### PR DESCRIPTION
**Pull request by stuart-rimel**
### :page_facing_up: Summary of Changes

Adds the time input component.

This component is pretty much just the native `<input type="time">` component with a couple of abstractions on top.
1. `seconds` prop as a boolean for controlling whether the seconds field is displayed within the component. This internally sets the `step` attribute to 1.
2. `timeOptions` prop which takes a list of time strings in the format `HH:mm`, `HH:mm:ss`. This internally sets the `list` attribute to point to an internally generated `<datalist>` element.

There are two types of dropdowns that come with the native `<input type="time">`, one is the default which allows users to select the time for each field. The other dropdown is the `<datalist>` that is supplied by either `timeOptions` or the `list` id of a `<datalist>` element on the same document. When both `timeOptions` and `list` are provided, the `<datalist>` which `list` points to will take precedence. [MDN reference](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#list) for more on using `list` with this input. 

This PR also removes the `autoFocus` prop from the date, number, select, and text inputs.

### :thought_balloon: Type of Change

- [ ] Bug fix (fixes an issue)
- [x] New feature (adds functionality)
- [ ] Configuration change (modifies scaling or properties of existing infrastructure & services)
- [ ] Documentation change
- [ ] Other

### :clipboard: Test Plan

Manually tested in storybook, included unit tests for 100% coverage.

### :white_check_mark: Self Code Review Checklist

PR authors and reviewers, please verify that all of these items have been completed.

- [x] My code is clean and readable
- [x] I followed standard conventions
- [x] Component code is WCAG 2.2 compliant (if applicable)
- [ ] I have made theme files changes (if applicable)
- [ ] I have made documentation changes (if applicable)
- [ ] I have made Storybook changes including usage documentation (if applicable)
- [ ] I have tested the component thoroughly in Storybook including RTL rendering (if applicable)

### :link: Azure DevOps Work Item

[AB#578576](https://dev.azure.com/ViewpointVSO/74194628-0b6e-4b03-95a4-2d06c069dcbe/_workitems/edit/578576)

### CODE REVIEW

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+  Event as StencilEvent,
+  EventEmitter,
+  h,
+  Host,
+  Prop,
+} from '@stencil/core';
+import { convertPropsToClasses } from './modus-wc-time-input.tailwind';
+import { ModusSize } from '../../types';
+
+const INTERNAL_DATALIST_NAME = 'modus-wc-internal-time-options';
```
Comment (gabriel-piltzer-trimble) [inline src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: What's the value of having this constant outside of the component?

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+   * A value is required for the form to be submittable.
+   */
+  @Prop() required?: boolean = false;
+
+  /**
+   * Displays the time input format as `HH:mm:ss` if `true`.
+   * Internally sets the `step` to 1 second.
+   * If a `step` value is provided, it will override this attribute.
+   */
+  @Prop() seconds?: boolean = false;
```
Comment (gabriel-piltzer-trimble) [inline src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: The naming on this isn't shouting `boolean`. Did you consider other names? Initially I thought `seconds` would be of type `number`.

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+   * Specifies the granularity that the `value` must adhere to.
+   * Value of step given in seconds. Default value is 60 seconds.
+   * Overrides the `seconds` attribute if both are provided.
+   */
+  @Prop() step?: number;
+
+  /**
+   * The options to display in the time input dropdown. Time options must be in `HH:mm` or `HH:mm:ss` format.
+   */
+  @Prop() timeOptions: string[] = [];
```
Comment (gabriel-piltzer-trimble) [inline src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: Would `datalistOptions` make more sense to align with HTML naming?

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+  /**
+   * Determine the control's relative ordering for sequential focus navigation (typically with the Tab key).
+   */
+  @Prop() inputTabIndex?: number;
+
+  /**
+   * ID of a <datalist> element that contains pre-defined time options.
+   * The value must be the ID of a <datalist> element in the same document.
+   */
+  @Prop({ mutable: true }) list?: string;
```
Comment (gabriel-piltzer-trimble) [inline src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: Would `datalistId` make more sense based on the comment?

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+  private handleInput = (event: Event) => {
+    this.inputChange.emit(event);
+  };
+
+  /**
+   * The ID of the internal <datalist> element. Unique to each instance of the time input component.
+   * This is used as the `list` id when `timeOptions` are provided.
+   */
+  private readonly internalDatalistId = `${INTERNAL_DATALIST_NAME}-${Math.random().toString(36).substring(2, 11)}`;
```
Comment (gabriel-piltzer-trimble) [inline src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: I recommend using `crypto.randomUUID()` or `nanoid` over custom logic

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+  private handleFocus = (event: FocusEvent) => {
+    this.inputFocus.emit(event);
+  };
+
+  private handleInput = (event: Event) => {
+    this.inputChange.emit(event);
+  };
+
+  /**
```
Comment (gabriel-piltzer-trimble) [inline src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: Be consistent with comment formatting. For internal object use `//`

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.scss [hunk 7]
```diff
@@ -0,0 +1,42 @@
+/**
+* This component uses Tailwind CSS and DaisyUI.
+* Only add styles here that should not be applied by Tailwind, Daisy, or the theme.
+*/
+
+[data-theme='modus-classic-dark'] .modus-wc-input {
+  color-scheme: dark; // needed to ensure the clock icon button is visible in dark mode
+}
+
+.modus-wc-input {
```
Comment (gabriel-piltzer-trimble) [inline src/components/atoms/modus-wc-time-input/modus-wc-time-input.scss]: This styling block should include both `data-theme` selectors, when not using these themes the components should look like DaisyUI (not Modus).

**Review by gabriel-piltzer-trimble — CHANGES_REQUESTED** [review]

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+          list={this.list}
+          max={this.max}
+          min={this.min}
+          name={this.name}
+          onBlur={this.handleBlur}
+          onFocus={this.handleFocus}
+          onInput={this.handleInput}
+          readonly={this.readOnly}
+          required={this.required}
+          step={this.step || !!this.seconds ? 1 : 60}
```
Comment (gabriel-piltzer-trimble) [inline src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]:
```suggestion
          step={this.step || this.seconds ? 1 : 60}
```

Does this work?

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx (hunk 6 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: fixed

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx (hunk 5 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: Great suggestions! I really like `nanoid` so I went that direction which led to me to adding `nanoid` as a dependency. See updated PR.

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx (hunk 4 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: Love it, renamed in updated PR

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx (hunk 3 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: I like that, fixed in updated PR

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx (hunk 2 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: Good point! I renamed this prop to be `showSeconds` in latest change set. Lmk what you think

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx (hunk 1 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: Definitely awkward. I refactored this name to be inlined during `insternalDatalistId` instantiation.
```ts
private readonly internalDatalistId = `${'modus-wc-datalist-id'}-${nanoid(10)}`;
```
Happy to refactor again if you'd prefer to still use a constant

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.scss (hunk 7 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.scss]: Nice catch, fixed in Updated PR

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx (hunk 8 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: fixed in updated PR

File: src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx (hunk 5 above)
Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: UPDATE: 

@gabriel-piltzer-trimble turns out I ran into major issues with jest and nanoid ESM modules. I tried configuring  `transformIgnorePatterns` in our stencil config, but still encountered the same unexpected token error from jest. I also tried mocking the module in a setup test file but got the same error back again. I'm only able to get jest to work with v3 of nanoid (they're on v5 currently). 

Going to try this again using `crypto.randomUUID()`

Comment (stuart-rimel) [reply src/components/atoms/modus-wc-time-input/modus-wc-time-input.tsx]: UPDATE 2:

Per our conversation today I've reverted this back to using the Math.random method but it's now in a utils file for generic use amongst our components

File: angular-value-accessor-bindings.ts [hunk 9]
```diff
@@ -14,6 +14,7 @@ const angularValueAccessorBindings: ValueAccessorConfig[] = [
       'modus-wc-number-input',
       'modus-wc-select',
       'modus-wc-text-input',
+      'modus-wc-time-input',
```
Comment (gabriel-piltzer-trimble) [inline angular-value-accessor-bindings.ts]: Alpha sort

Comment (stuart-rimel) [reply angular-value-accessor-bindings.ts]: fixed

**Review by gabriel-piltzer-trimble — APPROVED** [review]

**Pull request merged**